File: scripts/storage_service.py
import os

from abc import ABC, abstractmethod
from enum import Enum

debug: bool = False


class StorageType(Enum):
    FILE = "file"
    DATABASE = "database"


class StorageService(ABC):
    @abstractmethod
    def store(self, data: bytes, filepath: str) -> ():
        pass

    @abstractmethod
    def load(self, filepath: str) -> bytes:
        pass


class StorageServiceFactory:
    @classmethod
    def get_storage_service(cls, storage_type: StorageType) -> StorageService:
        if storage_type == StorageType.FILE:
            return FileStorageService()
        elif storage_type == StorageType.DATABASE:
            return DatabaseStorageService()
        else:
            raise ValueError(f"Unknown storage type: {storage_type}")


class FileStorageService(StorageService):
    def store(self, data: bytes, filepath: str) -> ():
        try:
            with open(filepath, 'wb') as f:
                f.write(data)
        except Exception as e:
            raise IOError(f"Error writing key: {e}")

    def load(self, filepath: str) -> bytes:
        try:
            with open(filepath, 'rb') as f:
                data = f.read()
        except Exception as e:
            raise IOError(f"Error reading data: {e}.")
        return data


class DatabaseStorageService(StorageService):
    def store(self, data: bytes, filepath: str) -> ():
        # Implement database storage logic here
        pass

    def load(self, filepath: str) -> bytes:
        # Implement database retrieval logic here
        pass


if __name__ == "__main__":
    storage_service = StorageServiceFactory.get_storage_service(StorageType.FILE)

    data = b"test_data"
    filepath = "temp_data.txt"

    storage_service.store(data, filepath)
    data_read = storage_service.load(filepath)

    if debug:
        print(data)
        print(data_read)

    assert data == data_read

    os.remove(filepath)

    print("All tests passed!")

File: scripts/storageService.js
const fs = require('fs');

const debug = false;

class StorageType {
    static FILE = "file";
    static DATABASE = "database";
}

class StorageService {
    store(data, filepath) {
        try {
            fs.writeFileSync(filepath, data);
        } catch (error) {
            throw new Error(`Error writing data: ${error}`);
        }
    }

    load(filepath) {
        try {
            const data = fs.readFileSync(filepath);
            return data;
        } catch (error) {
            throw new Error(`Error reading data: ${error}`);
        }
    }
}

class StorageServiceFactory {
    static getStorageService(storageType) {
        if (storageType === StorageType.FILE) {
            return new FileStorageService();
        } else if (storageType === StorageType.DATABASE) {
            return new DatabaseStorageService();
        } else {
            throw new Error(`Unknown storage type: ${storageType}`);
        }
    }
}

class FileStorageService extends StorageService {
    store(data, filepath) {
        try {
            fs.writeFileSync(filepath, data);
        } catch (error) {
            throw new Error(`Error writing data: ${error}`);
        }
    }

    load(filepath) {
        try {
            const data = fs.readFileSync(filepath);
            return data;
        } catch (error) {
            throw new Error(`Error reading data: ${error}`);
        }
    }
}

class DatabaseStorageService extends StorageService {
    store(data, filepath) {
        // Implement database storage logic here
    }

    load(filepath) {
        // Implement database retrieval logic here
    }
}

module.exports = {
    StorageType,
    StorageServiceFactory,
};

if (require.main === module) {
    const storageService = StorageServiceFactory.getStorageService(StorageType.FILE);

    const data = Buffer.from("test_data");
    const filepath = "temp_data.txt";

    storageService.store(data, filepath);
    const dataRead = storageService.load(filepath);

    if (debug) {
        console.log(data.toString());
        console.log(dataRead.toString());
    }

    console.assert(data.equals(dataRead));

    fs.unlinkSync(filepath);

    console.log("All tests passed!");
}
